packages/api/src/controllers/cdn-data.ts

```diff
@@ -0,0 +1,272 @@
+      if (!user) {
+        logger.error(`Can't find user for playbackId=${playbackId}`);
+        return null;
+      }
+      usersCache.set(playbackId, user);
+      return user;
+    };
+
+    const start = Date.now();
+    const dataAr = req.body as Array<SendData>;
+    console.log("Got data: ", JSON.stringify(dataAr, null, 2));
+    if (!Array.isArray(dataAr)) {
+      res.status(400);
+      res.end();
+    }
+    for (const data of dataAr) {
+      const hour = new Date(data.date * 1000).toUTCString();
+      if (!data.data?.length) {
+        continue;
+      }
+      // aggregate stream_ids into playback_ids
+      for (const row of data.data) {
+        // console.log(`===> checking row `, row)
+        if (row.playback_id) {
+          const user = await getUser(row.playback_id);
+          if (user) {
+            row.user_id = user.id;
+            row.user_email = user.email;
+          }
+        }
+        if (row.stream_id) {
+          // find playbackId by streamId
+          let stream = await db.stream.get(row.stream_id);
+          if (!stream) {
+            logger.error(`Invalid stream_id=${row.stream_id} in hour=${hour}`);
+            continue;
+          }
+          if (!stream.playbackId && stream.parentId) {
+            stream = await db.stream.get(stream.parentId);
+          }
+          if (!stream) {
+            logger.error(
+              `Invalid stream_id=${stream.parentId} in hour=${hour}`
+            );
+            continue;
+          }
+          if (!stream.playbackId) {
+            logger.error(
+              `For stream_id=${stream.parentId} in hour=${hour} can't find playbackId.`
+            );
+            continue;
+          }
+          const playbackRow = data.data.find(
+            (row) => row.playback_id === stream.playbackId
+          );
+          if (playbackRow) {
+            playbackRow.count += row.count;
+            playbackRow.total_cs_bytes += row.total_cs_bytes;
+            playbackRow.total_sc_bytes += row.total_sc_bytes;
+            playbackRow.total_filesize += row.total_filesize;
+            playbackRow.unique_users += row.unique_users;
+          } else {
+            row.playback_id = stream.playbackId;
+            row.stream_id = null;
+            const user = await getUser(row.playback_id);
+            if (user) {
+              row.user_id = user.id;
+              row.user_email = user.email;
+            }
+          }
+        }
+      }
+      // insert data into db
+      // @ts-ignore
+      const rows = data.data.filter((obj) => obj.playback_id && obj.user_id);
+      const badRows = data.data.filter((obj) => !obj.playback_id);
+      console.log(`==> bad rows:`, JSON.stringify(badRows, null, 2));
+      const badRows2 = data.data.filter((obj) => !obj.user_id);
+      console.log(`==> bad rows 2:`, JSON.stringify(badRows2, null, 2));
+      console.log(`==> good rows:`, JSON.stringify(rows, null, 2));
```

Comment:
Probably better to remove these logs in production (at least the "good rows" one, which should be a lot of data I guess)